docs/proposals/hpa/cronfederatedhpa.md

```diff
@@ -0,0 +1,339 @@
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec is the specification of the CronFederatedHPASpec.
+	// +required
+	Sepc CronFederatedHPASpec `json:"spec"`
+
+	// Status is the current status of the CronFederatedHPA.
+	// +optional
+	Status CronFederatedHPAStatus `json:"status"`
+}
+
+type CronFederatedHPASpec struct {
+	// ScaleTargetRef points to the target resource to scale.
+	// It could be FederatedHPA, Deployment, StatefulSet, or other resource with scale sub-resource.
+	// +required
+	ScaleTargetRef ScaleTargetRef `json:"scaleTargetRef"`
+
+	// Rules is the list of CronFederatedHPARule,
+	// it define when and how to scale the related resources.
+	Rules []CronFederatedHPARule `json:"rules"`
+}
+
+type CronFederatedHPARule struct {
+	// RuleName is the name of the CronFederatedHPARule.
+	// +required
+	RuleName string `json:"ruleName"`
+
+	// MaxDelaySeconds is the max delay of the CronFederatedHPARule.
+	// If the rule is not executed(like component crashed) between a time window, it's not necessary to do it again.
+	// +optional
+	MaxDelaySeconds *int32 `json:"maxDelaySeconds,omitempty"`
+
+	// TimeZone for the giving rule, if not specified, the system time zone will be used.
+	// see https://en.wikipedia.org/wiki/List_of_tz_database_time_zones for the all timezones.
+	// +optional
+	TimeZone *string
+
+	// Suspend is used to suspend the CronFederatedHPARule.
+	// The suspended rule will not be executed.
+	// Default is false
+	// +optional
+	Suspend *bool `json:"suspend"`
+
+	// Schedule is the schedule of the CronFederatedHPARule.
+	// The format should follow https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#schedule-syntax
+	// +required
+	Schedule string `json:"schedule"`
+
+	// TargetReplicas is the target replicas to be scaled for deployment, statefulset, etc.
+	// It should be set if you want to scale the related resources.
+	// +optional
+	TargetReplicas *int32 `json:"targetReplicas,omitempty"`
+
+	// TargetMinReplicas is the target MinReplicas to be set for FederatedHPA.
+	// It should be set if you want to configure FederatedHPA.
+	// If you not set this field, the value will not be updated.
+	// +optional
+	TargetMinReplicas *int32 `json:"targetMinReplicas,omitempty"`
+
+	// TargetMaxReplicas is the target MaxReplicas to be set for FederatedHPA.
+	// It should be set if you want to configure FederatedHPA.
+	// If you not set this field, the value will not be updated.
+	// +optional
+	TargetMaxReplicas *int32 `json:"targetMaxReplicas,omitempty"`
+}
+
+type ScaleTargetRef struct {
+	// APIVersion of the referent resources.
+	// +required
+	ApiVersion string `json:"apiVersion"`
+
+	// Kind of the referent resources.
+	// +required
+	Kind string `json:"kind"`
+
+	// Name of the referent resources.
+	// +required
+	Name string `json:"name"`
+}
+
+type CronFederatedHPAStatus struct {
+	// Conditions is the current status of the CronFederatedHPA.
+	RuleConditions []RuleCondition `json:"conditions"`
```

Comment:
It is more like rule execution history, not a condition. 

I tend to maintain a history for each rule rather than a condition for each rule.